tests/unit/evaluation/CrossValidation_multithread_unittest.cc

```diff
@@ -0,0 +1,119 @@
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * The views and conclusions contained in the software and documentation are those
+ * of the authors and should not be interpreted as representing official policies,
+ * either expressed or implied, of the Shogun Development Team.
+ *
+ */
+
+#include <shogun/base/init.h>
+#include <shogun/features/DenseFeatures.h>
+#include <shogun/labels/BinaryLabels.h>
+#include <shogun/kernel/GaussianKernel.h>
+#include <shogun/classifier/svm/LibSVM.h>
+#include <shogun/evaluation/CrossValidation.h>
+#include <shogun/evaluation/StratifiedCrossValidationSplitting.h>
+#include <shogun/evaluation/ContingencyTableEvaluation.h>
+#include <gtest/gtest.h>
+
+using namespace shogun;
+
+void generate_data(SGMatrix<float64_t>& mat, SGVector<float64_t> &lab)
+{
+	int32_t num=lab.size();
+	SG_SPRINT("%d", num);
+
+	for (index_t i=0; i<num; i++)
+	{
+		mat(0,i)=i<num/2 ? 0 : 100;
+		mat(1,i)=i;
+	}
+
+	for (index_t i=0; i<num; ++i)
+		lab.vector[i]=i<num/2 ? -1.0 : 1.0;
+
+}
+
+TEST(CrossValidation_multithread, LibSVM_unlocked)
+{
+	int32_t num=100;
+	SGMatrix<float64_t> mat(2, num);
+
+	/* training labels +/- 1 for each cluster */
+	SGVector<float64_t> lab(num);
+
+	/*create simple linearly separable data*/
+	generate_data(mat, lab);
+
+	CBinaryLabels* labels=new CBinaryLabels(lab);
+
+	CDenseFeatures<float64_t>* features=
+			new CDenseFeatures<float64_t>(mat);
+	SG_REF(features);	
+
+	int32_t kernel_cache=100;
+	int32_t width=10;
+	CGaussianKernel* kernel=new CGaussianKernel(kernel_cache, width);
+	kernel->init(features, features);
+
+	/* create svm via libsvm */
+	float64_t svm_C=10;
+	float64_t svm_eps=0.0001;
+	CLibSVM* svm=new CLibSVM(svm_C, kernel, labels);
+	svm->set_epsilon(svm_eps);
+
+	/* evaluation criterion */
+	CContingencyTableEvaluation* eval_crit=
+			new CContingencyTableEvaluation(ACCURACY);
+
+	/* splitting strategy */
+	index_t n_folds=4;
+	CStratifiedCrossValidationSplitting* splitting=
+			new CStratifiedCrossValidationSplitting(labels, n_folds);
+
+	CCrossValidation* cross=new CCrossValidation(svm, features, labels,
+			splitting, eval_crit);
+
+	cross->set_autolock(false);
+	cross->set_num_runs(4);
+	cross->parallel->set_num_threads(1);	
+
+	CCrossValidationResult* result1=(CCrossValidationResult*)cross->evaluate();
+	float64_t mean1 = result1->mean;
+	
+	cross->parallel->set_num_threads(cross->parallel->get_num_cpus());
+
+	CCrossValidationResult* result2=(CCrossValidationResult*)cross->evaluate();
+	float64_t mean2 = result2->mean;
+
+	EXPECT_EQ(mean1, mean2);
+
```

Comment:
Cool -- if it is clean. Then please polish the PR and we can go on.

Before I merge, Can you put some benchmark numbers?